File: insertion/insertion_sort.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func insertionSort(arr []int, len int) []int {
	for j := 1; j < len; j++ {
		key := arr[j]

		i := j - 1
		for i >= 0 && arr[i] > key {
			arr[i+1] = arr[i]
			i = i - 1
		}

		arr[i+1] = key
	}

	return arr
}

func main() {
	rand.Seed(time.Now().UnixNano())
	arr := make([]int, 10000)

	for i := 0; i < len(arr); i++ {
		for {
			val := rand.Intn(10000)
			isDuplicate := false

			for j := 0; j < i; j++ {
				if val == arr[j] {
					isDuplicate = true
					break
				}
			}

			if !isDuplicate {
				arr[i] = val
				break
			}
		}
	}

	// fmt.Println(arr)
	s := 0
	for i := 0; i < 10; i++ {
		carr := make([]int, 10000)
		copy(carr, arr)

		start := time.Now()
		insertionSort(carr, 10000)
		fmt.Println("Excution Time: ", time.Since(start))

		fmt.Println(carr)

		break
	}
	fmt.Println(s)
}

File: insertion-c/insertion_sort.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>

int* insertionSort(int* arr, int len) {
    for(int j = 2; j < len; j++) {
        int key = arr[j];
        int i = j - 1;

        while(i >= 0 && arr[i] > key) {
            arr[i+1] = arr[i];
            i -= 1;
        }

        arr[i+1] = key;
    }

    return arr;
}

int main() {
    int arr[10000];
    for(int i = 0; i < 10000; i++) {
        while(1) {
            int val = rand()%10000;
            int isDuplicated = 0;

            for(int j = 0; j < i; j++) {
                if(val == arr[j]) {
                    isDuplicated = 1;
                    break;
                }
            }

            if(!isDuplicated) {
                arr[i] = val;
                break;
            }
        }
    }

    for(int i = 0; i < 10; i++) {
        int carr[10000];
        memcpy(carr, arr, sizeof(arr));
        clock_t start = clock();
        insertionSort(carr, 10000);
        clock_t end = clock();
        printf("Execution Time: %f ms\n", ((double)(end - start) / CLOCKS_PER_SEC) * 1000);
    }

}